File: CS 3450 Code Examples/3450Code/Skeleton.java
abstract class Base {
    public final void theAlgorithm() {
        fixedop1();
        missingop1();
        fixedop2();
        missingop2();
    }
    final void fixedop1() {
        System.out.println("fixedop1");
    }
    final void fixedop2() {
        System.out.println("fixedop2");
    }
    abstract void missingop1();
    abstract void missingop2();
};

class Derived extends Base {
    void missingop1() {
        System.out.println("missingop1");
    }
    void missingop2() {
        System.out.println("missingop2");
    }
};

class Skeleton {
    public static void main(String[] args) {
        Derived d = new Derived();
        d.theAlgorithm();
    }
}

File: CS 3450 Code Examples/3450Code/skeleton.cpp
#include <iostream>
using namespace std;

class Base {
public:
    void theAlgorithm() {
        fixedop1();
        missingop1();
        fixedop2();
        missingop2();
    }
    void fixedop1() {
        cout << "fixedop1\n";
    }
    void fixedop2() {
        cout << "fixedop2\n";
    }
    virtual void missingop1() = 0;
    virtual void missingop2() = 0;
};

class Derived : public Base {
    void missingop1() {
        cout << "missingop1\n";
    }
    void missingop2() {
        cout << "missingop2\n";
    }
};

int main() {
    Derived d;
    d.theAlgorithm();
}
